File: Java/workflows/src/main/java/com/r3/developers/samples/CarRegistry/workflows/ListCarFlow.java
package com.r3.developers.samples.CarRegistry.workflows;


import com.r3.developers.samples.TokenisedCarRegistry.states.CarState;
import net.corda.v5.application.crypto.DigestService;
import net.corda.v5.application.flows.ClientRequestBody;
import net.corda.v5.application.flows.ClientStartableFlow;
import net.corda.v5.application.flows.CordaInject;
import net.corda.v5.application.marshalling.JsonMarshallingService;
import net.corda.v5.application.membership.MemberLookup;
import net.corda.v5.base.annotations.Suspendable;

import net.corda.v5.ledger.utxo.StateAndRef;
import net.corda.v5.ledger.utxo.UtxoLedgerService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


import java.util.List;
import java.util.stream.Collectors;

import static java.util.Objects.requireNonNull;

public class ListCarFlow implements ClientStartableFlow {

    private final static Logger log = LoggerFactory.getLogger(ListCarFlow.class);

    // Injects the JsonMarshallingService to read and populate JSON parameters.
    @CordaInject
    public JsonMarshallingService jsonMarshallingService;

    // Injects the UtxoLedgerService to enable the flow to make use of the Ledger API.
    @CordaInject
    public UtxoLedgerService utxoLedgerService;

    @CordaInject
    private DigestService digestService;

    @CordaInject
    private MemberLookup memberLookup;

    @Suspendable
    @Override
    public String call(ClientRequestBody requestBody) {

        log.info("LCF 1");

        // Queries the VNode's vault for unconsumed states and converts the result to a serializable DTO.
        List<StateAndRef<CarState>> states = utxoLedgerService.findUnconsumedStatesByType(CarState.class);
        log.info("LCF 2");
        List<ListCarFlowResults> results = states.stream().map(stateAndRef ->
                new ListCarFlowResults(
                        stateAndRef.getState().getContractState().getIdentifier(),
                        stateAndRef.getState().getContractState().getIssuerInfo(),
                        stateAndRef.getState().getContractState().getOwnerInfo(),
                        stateAndRef.getState().getContractState().getThirdPartyInfo(),
                        stateAndRef.getState().getContractState().getMileage(),
                        stateAndRef.getState().getContractState().isConsumed(),
                        stateAndRef.getState().getContractState().getReasonForRequest(),
                        stateAndRef.getState().getContractState().getLscTire(),
                        stateAndRef.getState().getContractState().getLscOil(),
                        stateAndRef.getState().getContractState().getLscCoolant(),
                        stateAndRef.getState().getContractState().getLscFilter(),
                        stateAndRef.getState().getContractState().getLscBattery(),
                        stateAndRef.getState().getContractState().getAmountOfTimesServiced(),
                        stateAndRef.getState().getContractState().getAmountOfRepairs(),
                        stateAndRef.getState().getContractState().getRepairReasons(),
                        stateAndRef.getState().getContractState().getMods(),
                        stateAndRef.getState().getContractState().getCurrentOwnerMileage(),
                        stateAndRef.getState().getContractState().getUserName(),
                        stateAndRef.getState().getContractState().getMakeAndModel()
                )
        ).collect(Collectors.toList());
        log.info("LCF 3");
        // Uses the JsonMarshallingService's format() function to serialize the DTO to Json.
        return jsonMarshallingService.format(results);
    }
}
/*
RequestBody for triggering the flow via http-rpc:
{
    "clientRequestId": "list-1",
    "flowClassName": "com.r3.developers.samples.CarRegistry.workflows.ListCarFlow",
    "requestBody": {}
}
*/

File: Kotlin/workflows/src/main/kotlin/com/r3/developers/samples/CarRegistry/workflows/ListCarFlow.kt
package com.r3.developers.samples.CarRegistry.workflows

import com.r3.developers.samples.CarRegistry.states.CarState
import net.corda.v5.application.flows.ClientRequestBody
import net.corda.v5.application.flows.ClientStartableFlow
import net.corda.v5.application.flows.CordaInject
import net.corda.v5.application.marshalling.JsonMarshallingService
import net.corda.v5.base.annotations.Suspendable
import net.corda.v5.ledger.utxo.UtxoLedgerService
import org.slf4j.LoggerFactory
import java.util.stream.Collectors

// This flow is used to list all the gold tokens available in the vault.
class ListCarFlow: ClientStartableFlow {

    private companion object {
        val log = LoggerFactory.getLogger(this::class.java.enclosingClass)
    }

    @CordaInject
    lateinit var jsonMarshallingService: JsonMarshallingService

    // Injects the UtxoLedgerService to enable the flow to make use of the Ledger API.
    @CordaInject
    lateinit var utxoLedgerService: UtxoLedgerService

    @Suspendable
    override fun call(requestBody: ClientRequestBody): String {
        val states = utxoLedgerService.findUnconsumedStatesByType(
            CarState::class.java
        )

        // Queries the VNode's vault for unconsumed states and converts the result to a serializable DTO.
        val results = states.stream().map{
            CarStateList(
                it.state.contractState.ownerInfo    ,
                it.state.contractState.userName,
                it.state.contractState.makeAndModel,
                it.state.contractState.reasonForRequest,
                it.state.contractState.mileage,
                it.state.contractState.consumed,
                it.state.contractState.repairReasons,
                it.state.contractState.mods,
                it.state.contractState.currentOwnerMileage,
                it.state.contractState.lscTire,
                it.state.contractState.lscOil,
                it.state.contractState.lscCoolant,
                it.state.contractState.lscFilter,
                it.state.contractState.lscBattery,
                it.state.contractState.amountOfTimesServiced,
                it.state.contractState.identifier
            )
        }.collect(Collectors.toList())
        // Uses the JsonMarshallingService's format() function to serialize the DTO to Json.
        return jsonMarshallingService.format(results)
    }
}

data class CarStateList(
    val ownerInfo: String,
    val userName: String,
    val makeAndModel: String,
    val reasonForRequest: String,
    val mileage : Int,
    val consumed : Boolean,
    val repairReasons : List<String>,
    val mods : List<String>,
    val currentOwnerMileage : Int,
    val lscTire: Int,
    val lscOil: Int,
    val lscCoolant: Int,
    val lscFilter: Int,
    val lscBattery: Int,
    val amountOfTimesServiced: Int,
    val identifier: Int
)